pallets/sto/src/lib.rs

```diff
@@ -0,0 +1,353 @@
+// Copyright (c) 2020 Polymath
+
+//! # STO Module
+//!
+//! This is a proof of concept module. It is not meant to be used in the real world in its' current state.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+#![recursion_limit = "256"]
+
+use codec::{Decode, Encode};
+use frame_support::{
+    decl_error, decl_event, decl_module, decl_storage, dispatch::DispatchResult, ensure,
+};
+use frame_system::ensure_signed;
+use pallet_identity as identity;
+use pallet_portfolio::{PortfolioAssetBalances, Trait as PortfolioTrait};
+use pallet_settlement::{
+    self as settlement, Leg, ReceiptDetails, SettlementType, Trait as SettlementTrait, VenueInfo,
+    VenueType,
+};
+use pallet_timestamp::{self as timestamp, Trait as TimestampTrait};
+use polymesh_common_utilities::{
+    traits::{asset::Trait as AssetTrait, identity::Trait as IdentityTrait},
+    with_transaction, CommonTrait, Context,
+};
+use polymesh_primitives::{IdentityId, PortfolioId, Ticker};
+use sp_runtime::traits::{CheckedAdd, CheckedMul};
+use sp_std::prelude::*;
+
+type Identity<T> = identity::Module<T>;
+type Settlement<T> = settlement::Module<T>;
+type Timestamp<T> = timestamp::Module<T>;
+
+/// Details about the Fundraiser
+#[cfg_attr(feature = "std", derive(Debug))]
+#[derive(Encode, Decode, Default, Clone, PartialEq, Eq, PartialOrd, Ord)]
+pub struct Fundraiser<Balance, Moment> {
+    /// Portfolio containing the asset being offered
+    pub offering_portfolio: PortfolioId,
+    /// Asset being offered
+    pub offering_asset: Ticker,
+    /// Portfolio receiving funds raised
+    pub raising_portfolio: PortfolioId,
+    /// Asset to receive payment in
+    pub raising_asset: Ticker,
+    /// Tiers of the fundraiser.
+    /// Each tier has a set amount of tokens available at a fixed price.
+    /// The sum of the tiers is the total amount available in this fundraiser.
+    pub tiers: Vec<FundraiserTier<Balance>>,
+    /// Id of the venue to use for this fundraise
+    pub venue_id: u64,
+    /// Start of the fundraiser
+    pub start: Moment,
+    /// End of the fundraiser
+    pub end: Option<Moment>,
+    /// Fundraiser is frozen
+    pub frozen: bool,
+}
+
+/// Single tier of a tiered pricing model
+#[cfg_attr(feature = "std", derive(Debug))]
+#[derive(Encode, Decode, Default, Clone, PartialEq, Eq, PartialOrd, Ord)]
+pub struct PriceTier<Balance> {
+    total: Balance,
+    price: Balance,
+}
+
+/// Single price tier of a `Fundraiser`.
+/// Similar to a `PriceTier` but with an extra field `remaining` for tracking the amount available for purchase in a tier.
+#[cfg_attr(feature = "std", derive(Debug))]
+#[derive(Encode, Decode, Default, Clone, PartialEq, Eq, PartialOrd, Ord)]
+pub struct FundraiserTier<Balance> {
+    inner: PriceTier<Balance>,
```

Comment:
It might be better to not nest the `PriceTier` like this. We can have `total`, `price` and `remaining` at the same level.